Include the Dirac and Polar Fourier basis Classes

junchaoxia opened the pull request:
This PR will include the Dirac and Polar Fourier basis Classes, as parts of  work to address the issue #109,  integrate the orientation method of synchronization from the old python code into the new frame work.

junchaoxia committed:
Implement the derived class for Dirac basis

junchaoxia committed:
Initially add the class for 2D polar Fourier basis function

janden commented on src/aspire/basis/dirac.py:
```diff
@@ -1,5 +1,79 @@
+    """
+    def __init__(self, sz, mask=None):
+        """
+        initialize an object for Dirac Basis
```
Capitalize “initialize” and do not capitalize “basis”

junchaoxia replied:
Done.

janden commented on src/aspire/basis/dirac.py:
```diff
@@ -1,5 +1,79 @@
+    def __init__(self, sz, mask=None):
+        """
+        initialize an object for Dirac Basis
+        :param sz: The size of the vectors for which to define the basis.
```
“Shape”. “Size” is a MATLAB terminology. Probably replace `sz` with `shape`  as well.

junchaoxia replied:
Done. The `sz` parameter was used for all of basis class. I will keep it now. Probably I should refactor all of them later.

janden commented on src/aspire/basis/dirac.py:
```diff
@@ -1,5 +1,79 @@
+        initialize an object for Dirac Basis
+        :param sz: The size of the vectors for which to define the basis.
+        :param mask: A boolean mask of size sz indicating which coordinates
+            to include in the basis (default true(sz)).
```
`true(sz)` → `np.full(sz, True)`

junchaoxia replied:
Done.

janden commented on src/aspire/basis/dirac.py:
```diff
@@ -1,5 +1,79 @@
+            to include in the basis (default true(sz)).
+        """
+        if mask is None:
+            mask = np.ones(sz)
```
`mask = np.full(sz, True)`.

junchaoxia replied:
Done.

janden commented on src/aspire/basis/dirac.py:
```diff
@@ -1,5 +1,79 @@
+
+        self.mask = m_flatten(mask)
+        self.count = np.sum(mask)
+        self.sz_prod = self.nres**self.ndim
```
Spaces around `**`. Also should `sz_prod` be public?

junchaoxia replied:
Done.

janden commented on src/aspire/basis/dirac.py:
```diff
@@ -1,5 +1,79 @@
+            dimensions correspond to dimensions two and higher of `v`.
+        """
+        v, sz_roll = unroll_dim(v, 2)
+        x = np.zeros(shape=tuple([self.sz_prod] + list(v.shape[1:])))
```
Isn't `shape=(self.sz_prod,) + v.shape[1:]` simpler? Same below.

junchaoxia replied:
Done.

janden commented on src/aspire/basis/dirac.py:
```diff
@@ -1,5 +1,79 @@
+
+        super().__init__(sz)
+
+        self.mask = m_flatten(mask)
```
Maybe make this private since it doesn't correspond to the `mask` argument given in the constructor?

junchaoxia replied:
Done.

janden commented on src/aspire/basis/dirac.py:
```diff
@@ -1,5 +1,79 @@
+        :param mask: A boolean mask of size sz indicating which coordinates
+            to include in the basis (default true(sz)).
+        """
+        if mask is None:
```
Call constructor of superclass?

junchaoxia replied:
It does call superclass. Not sure there is any addition.

janden replied:
It's generally best to call it first thing in the constructor so it doesn't overwrite anything.

junchaoxia replied:
For other basis classes, the base constructor is designed to be called at the end of the derived constructor because the `_build` function is included in the base constructor. I will follow the consistent way if there is no specific better reason.

janden replied:
Do you remember why we did that? It seems like it could cause problems.

junchaoxia replied:
If the base constructor is called first, that  `_build` will be called before the parameters specified in the derived class are initialized. This is not what we want, we want all parameters are initialized first then call  `_build` to initailize more complex data structure.

janden commented on src/aspire/basis/dirac.py:
```diff
@@ -1,5 +1,79 @@
+    def expand(self, x):
+        return self.evaluate_t(x)
+
+    def expand_t(self, x):
```
Remove. We're no longer implementing `expand_t`. It is ill-defined.

junchaoxia replied:
Removed.

janden commented on tests/test_Diracbasis.py:
```diff
@@ -0,0 +1,30 @@
+        pass
+
+    def testDiracEvaluate(self):
+        v = np.load(os.path.join(DATA_DIR, 'fbbasis_coefficients_8_8.npy'))
```
Can't we just generate this data on the fly? We should try to make these tests as self-contained as possible.

junchaoxia replied:
I tried to reuse the test data from Fast FB class so that we will not include new binary data. In the new commit, I built the array in fly.

janden commented on src/aspire/basis/polar_2d.py:
```diff
@@ -0,0 +1,111 @@
+
+    def _precomp(self):
+        """
+        Precomute the basis functions on a polar Fourier grid
```
These aren't actually basis functions, just grid points.

junchaoxia replied:
Yes. You are right.  Changed the description here and other places.

janden commented on src/aspire/basis/polar_2d.py:
```diff
@@ -0,0 +1,111 @@
+
+        :param x: The coefficient array in the standard 2D coordinate basis to be
+            evaluated. The first two dimensions must equal `self.sz`.
+        :return v: The evaluation of the coefficient array `v` in the polar Fourier basis.
```
We should have the radius and angle only occupy a single axis, just like the other basis objects.

junchaoxia replied:
I noticed this before. I kept this way based on the old Python code. Will change to a single axis after debugging the orientation estimation code.

janden commented on src/aspire/basis/polar_2d.py:
```diff
@@ -0,0 +1,111 @@
+        freqs *= omega0
+        return freqs
+
+    def evaluate_t(self, x):
```
Why don't we have `evaluate`? It should be fairly straightforward to compute the dual of `evaluate_t`.

junchaoxia replied:
Yes. I thought this before. There is no implementation of evaluate in the original Matlab code. It will be great if I can be shown the exact equation with the right normalization factor.

junchaoxia replied:
Done.

janden commented on src/aspire/basis/polar_2d.py:
```diff
@@ -0,0 +1,111 @@
+            This is an array of vectors whose first two dimensions are `self.nrad` and
+            `self.ntheta` and whose remaining dimensions correspond to higher dimensions of `x`.
+        """
+        return self.evaluate_t(x)
```
This isn't really valid. The polar Fourier transform is not an orthogonal basis, so the inverse doesn't equal the transpose.

junchaoxia replied:
For simplicity, I will remove it.

junchaoxia replied:
removed.

janden commented on src/aspire/basis/polar_2d.py:
```diff
@@ -0,0 +1,111 @@
+        # finufftpy require it to be aligned in fortran order
+        half_size = self.nrad * self.ntheta // 2
+        pf = np.empty((half_size, nimgs), dtype='complex128', order='F')
+        finufftpy.nufft2d2many(self.freqs[0, 0:half_size], self.freqs[1, 0:half_size], pf, 1, 1e-15, x)
```
`0:half_size` → `:half_size`

junchaoxia replied:
Done.

janden commented on src/aspire/basis/polar_2d.py:
```diff
@@ -0,0 +1,111 @@
+        pf = np.empty((half_size, nimgs), dtype='complex128', order='F')
+        finufftpy.nufft2d2many(self.freqs[0, 0:half_size], self.freqs[1, 0:half_size], pf, 1, 1e-15, x)
+        pf = m_reshape(pf, (self.nrad, self.ntheta // 2, nimgs))
+        v = np.concatenate((pf, pf.conj()), axis=1).copy()
```
Why `copy`?

junchaoxia replied:
Not necessary. Removed.

janden commented on tests/test_PolarBasis2D.py:
```diff
@@ -0,0 +1,21 @@
+        pass
+
+    def testPolarBasis2DEvaluate_t(self):
+        x = np.load(os.path.join(DATA_DIR, 'fbbasis_coefficients_8_8.npy'))
```
Same here. Is there no way to avoid hard-coded test data here? It's a relatively simple basis decomposition, so we could calculate it explicitly for a couple of values.

junchaoxia replied:
I removed the binary file and included the data directly.

janden reviewed (changes requested):
Looks good overall. Some things to discuss.

junchaoxia committed:
Improve and refactor DiracBasis class

junchaoxia committed:
Improve and refactor PolarBasis2D class

junchaoxia committed:
Add unit test for evaluate function of PolarFourier class

janden commented on src/aspire/basis/dirac.py:
```diff
@@ -1,5 +1,75 @@
+        self.count = np.sum(mask)
+        self._sz_prod = self.nres ** self.ndim
+
+    def _build(self):
```
Shouldn't this be called at the end of the constructor?

junchaoxia replied:
The `_build` function is called at the end of constructor of base class. In this derived class  `_build` is overrode and called at the end of constructor too.  Some rearranges were made in the new commit dfb2805.

janden commented on src/aspire/basis/polar_2d.py:
```diff
@@ -0,0 +1,122 @@
+            coordinate basis. This is an array whose first two dimensions equal `self.sz`
+            and the remaining dimensions correspond to dimensions two and higher of `v`.
+        """
+        # TODO: need check the normalization factor and develop unit test
```
Why are we missing a normalization factor?

junchaoxia replied:
The `evaluate_t` corresponding the `cryo_pft`  function in the Matlab code. But `evaluate`  is not defined. I am not sure the normalization factors for both of them are right to satisfy x=evaluate(evaluate_t(x)).

junchaoxia replied:
I added the normalization factor based on the equations from the website, http://fourier.eng.hmc.edu/e101/lectures/Image_Processing/node6.html
It has no effects on common line and orientation estimation but does have significant effects on satisfying x=evaluate(evaluate_t(x)).

junchaoxia replied:
The Normalization factor was removed the new commit fefb446.

janden commented on src/aspire/basis/polar_2d.py:
```diff
@@ -0,0 +1,122 @@
+            # try to use the same number as Fast FB basis
+            self.ntheta = 8 * self.nrad
+
+        super().__init__(size)
```
Again, call this first.

junchaoxia replied:
Same as other basis classes, the base constructor is designed to be called at the end of the derived constructor because the `_build` function is included in the base constructor.

janden commented on src/aspire/basis/polar_2d.py:
```diff
@@ -0,0 +1,122 @@
+
+        super().__init__(size)
+
+    def _build(self):
```
Again, shouldn't this be called at the end of the constructor?

junchaoxia replied:
It does as above.

janden commented on src/aspire/basis/polar_2d.py:
```diff
@@ -0,0 +1,122 @@
+        omega0 = 2 * np.pi / (2 * self.nrad - 1)
+        dtheta = 2 * np.pi / self.ntheta
+
+        # Note: The size of ntheta can be reduced by half if only evaluate_t function is needed as
```
Do we need this comment?

junchaoxia replied:
removed.

janden commented on src/aspire/basis/polar_2d.py:
```diff
@@ -0,0 +1,122 @@
+        """
+        # TODO: need check the normalization factor and develop unit test
+        v, sz_roll = unroll_dim(v, 3)
+        v = m_reshape(v, (self.nrad*self.ntheta, v.shape[2]))
```
Space around `*`.

junchaoxia replied:
Done.

janden commented on src/aspire/basis/polar_2d.py:
```diff
@@ -0,0 +1,122 @@
+        v = np.concatenate((pf, pf.conj()), axis=1)
+
+        # return v coefficients with two dimensions of (self.nrad, self.ntheta)
+        # TODO: Squeeze the first two dimensions into one for consistent with other basis classes
```
Why aren't we doing this already?

junchaoxia replied:
Done in the new commit.

janden commented on src/aspire/basis/polar_2d.py:
```diff
@@ -0,0 +1,122 @@
+        nimgs = v.shape[1]
+        # finufftpy require it to be aligned in fortran order
+        pf = np.empty((self._sz_prod, nimgs), dtype='complex128', order='F')
+        finufftpy.nufft2d1many(self.freqs[0], self.freqs[1], v, 1, 1e-15, self.sz[0], self.sz[1], pf)
```
Should be able to only give half of the `v` coefficients here (just like in `evaluate_t`).

junchaoxia replied:
I checked the FB2D basis classes. It looks that the other half is still needed for evaluating.

janden replied:
Even if you cast the output to real? Also, `pf` is not a polar Fourier transform here, so it needs a different name.

junchaoxia replied:
To satisfy the adjoint operator, I can not cast the output to real. So probably I will left it as it.  How about change `pf` to `x` as in other basis class?

janden replied:
> To satisfy the adjoint operator, I can not cast the output to real.

But `evaluate_t` only calculates the Fourier transform for half the frequencies, so its input must be real. Its adjoint must therefore have real outputs.

> How about change pf to x as in other basis class?

Yes that sounds good.

junchaoxia replied:
For `evaluate_t,` the input x vector is real, the the first half of output v vector is obtained from the FFT transform and the second half is obtained by the conjugate property. For the `evaluate,` the input v vector is complex, but the output x should be real. It looks that i can not just double the real value from the first half frequencies such as 
`finufftpy.nufft2d1many(self.freqs[0, :half_size], self.freqs[1, :half_size], v[:half_size, ...], 1, 1e-15, self.sz[0], self.sz[1], x)`.  In FB basis class, it looks that both - and + parts have to be evaluated.

garrettwrong replied:
Unrelated, but can I ask why we don't use the nufft class structure here? We do use it elsewhere.  This would allow the nufft calls can be automatically swapped out depending on the backend.  As this is written it will ignore user configuration selecting a different library etc, and also will not work in the case finufft is N/A.  Is this just for testing? Thanks

junchaoxia replied:
The implementation of passing of many frames is not in the develop branch yet. I did this way so that I should not repeat @garrettwrong's work for including the new functionality of finufft. We can easily to switch to when the new interface of many frames is ready to use.

garrettwrong replied:
_But this is an active request to put this code into the `development` branch?_

The nufft class was something Vineet wrote a long time ago.  While I appreciate saying you do not want to duplicate the `many` method work, you should know using many is only an optimization, it is not required to call finufft or the other nufft libraries that way.  Also, we both know your code was written long before my recent work _for the hackathon_ ;)

Many does perform better, but clearly since we ported code inside out, we're not really concerned with performance at this time...  Using the single transform setup, execution, and destructor does exist packaged as the master's nufft class, and is what we use elsewhere, like under the cov2d example.  It is actually the entire setup, execute, and destroy that I am concerned about, not the optimization.  To implement a many _optimization_ when using the class structure is a few line change later.  To survey, catalog, and re-implement code we just "ported" to use the ASPIRE python class structures, seems like a core concern to me.  I guess I don't understand the process we are using for new code.  Are we intending to rewrite it several times in the development branch?

As I see it, the problem with implementing everything adhoc again in the python code, is that someone else will need to come along and rewrite the ported codes (perhaps several times it seems) before they are production worthy... so it is repeating work, just in the future. That is why I asked if this was just for testing purposes.

Imagine some user/developer doesn't have finufft, or wants to use a different library, turns the ASPIRE configuration to a different library and none of this code works... bad times.

junchaoxia replied:
Yes. Many parts of code in the development branch can be reimplemented in a better way.  We are trying to reach a minimum set of modules using our new framework which can perform the whole pipeline of reconstruction. We would like to reach this in the summer but decide to postpone it to the end of 2020 for a reasonable estimation.  Currently the CPU version of finufft was installed through PyPI package. Of course we might change it later.  We can discuss these in the meeting.

janden replied:
> It looks that i can not just double the real value from the first half frequencies

I don't quite understand why this cannot be done? Taking the real part after performing the IFT corresponds to symmetrizing before the frequencies, then taking the IFT. I will take a stab.

Regarding the direct call to the Finufft module, part of this is my fault. To get some reasonable speeds for the covariance estimation, I started doing this instead of (properly) implementing the many interface in the Nufft wrapper classes. Why don't we switch to using the wrapper classes now (at a certain loss in performance) and prioritize getting those wrappers to use the many interface as soon as possible?

janden commented on tests/test_PolarBasis2D.py:
```diff
@@ -0,0 +1,127 @@
+        v[2, 16] = 1.0
+        v[2, 0] = 1.0
+        x = self.basis.evaluate(v)
+        result = np.array([
```
Where does this come from?

junchaoxia replied:
Try to see the result in coordinate space is reasonable or not just using two symmetric nonzero points in Polar Fourier space.

junchaoxia replied:
I recreated a new case for this function.

janden commented on tests/test_PolarBasis2D.py:
```diff
@@ -0,0 +1,127 @@
+from aspire.basis.polar_2d import PolarBasis2D
+
+
+class PolarBasis2DTestCase(TestCase):
```
Can we add a test to verify that `evaluate_t` is the adjoint of `evaluate`?

junchaoxia replied:
Do you mean that `evaluate_t` has the similar result of  `expand`?  Or x = evaluate_t(evaluate(x))?  I checked it yesterday, x != evaluate_t(evaluate(x)).

junchaoxia replied:
In higher reslution (larger nrad and ntheta), x looks closer to evaluate_t(evaluate(x)). Probably we need to use a disk mask, but this will not reproduce the results from previous code. I think it is OK for common line and orientation estimation but not accurate enough for image expanding.

junchaoxia replied:
in the new commit fefb446, a new function was added to test whether the `evaluate` is the adjoint operator of `evaluate_t` , namely satisfying  (y, A*x) = (A^t*y,x) = (B*y, x).

janden reviewed (changes requested):
Good. Some things to check.

junchaoxia committed:
Rearrange the constructor and _build function

junchaoxia committed:
Move expand function to base class

junchaoxia committed:
Change representing dimensions and add normalization factor in PolarFourier 2D class

junchaoxia committed:
Add unit test for ajoint operator and remove the normalization factor

junchaoxia committed:
fix typos in names

janden reviewed (changes requested):
One thing.

janden committed:
Update PF2D adjoint test

Must ensure the PF lies in the Hermitian symmetric subspace since we
expect images to be real.

janden committed:
Only use half of freqs in PolarBasis2D.evaluate

We can do this since images are real and their FT has Hermitian
symmetry.

janden committed:
Rename pf → x

junchaoxia commented on tests/test_PolarBasis2D.py:
```diff
@@ -213,6 +213,16 @@ def testPolarBasis2DAdjoint(self):
         # Namely, if A = evaluate, B = evaluate_t, and B=A^t, we will have
         # (y, A*x) = (A^t*y, x) = (B*y, x)
         x = np.random.randn(self.basis.count)
+
```
I think we do this transformation inside the `evaluate`  function. Should we remove this recalculation based on the conjugate part?

janden replied:
I'm not sure which transformation you're referring to. Line 219 below? We need this for the inner product `np.dot(y_t, x)` to be correct. Line 226 calls `evaluate`, which indeed symmetrizes `x` (just like line 219). However, when `x` is used in `np.dot(y_t, x)`, that `x` will not be symmetrized. The underlying issue is that `evaluate` does not operate on the entire space of polar Fourier coefficients, only on the subspace of Hermitian symmetric coefficients (since these correspond to real images). For the FB basis, we don't have this problem, because we explicitly break the symmetry by only storing half the frequencies. It might make sense to do this here as well, but currently the common lines assumes that all the frequencies are calculated, so that would break if we only give it half (although it's possible to rewrite it to only take half).

junchaoxia replied:
OK. I understand. If we use the output from `evalute_t` as `x,` then we don't need do this.

janden replied:
Correct, that would work as well.

junchaoxia replied:
Please approve the PR if you think this is the last issue. I will submit a new PR for the orientation code which depends on this one.
